mava/utils/jax_training_utils.py

```diff
@@ -46,42 +49,88 @@ def compute_running_mean_var_count(
     M2 = m_a + m_b + jnp.square(delta) * count * batch_count / tot_count
     new_var = M2 / tot_count
     new_count = tot_count
+    new_std = jnp.sqrt(new_var)
 
-    return jnp.array([new_mean, new_var, new_count])
+    return dict(mean=new_mean, var=new_var, std=new_std, count=new_count)
 
 
-def normalize(stats: jnp.ndarray, batch: jnp.ndarray) -> jnp.ndarray:
+def normalize(stats: Any, batch: jnp.ndarray) -> jnp.ndarray:
     """Normlaise batch of data using the running mean and variance.
 
-    stats (array) -- mean, var, count.
+    stats (Any)   -- dictionary with running mean, var, std, count.
     batch (array) -- current batch of data.
 
     Returns:
         denormalize batch (array)
     """
 
-    mean, var = stats[0], stats[1]
-    normalize_batch = (batch - mean) / (jnp.sqrt(jnp.clip(var, a_min=1e-2)))
+    mean, std = stats["mean"], stats["std"]
+    normalize_batch = (batch - mean) / (std + 1e-8)
 
     return normalize_batch
 
 
-def denormalize(stats: jnp.ndarray, batch: jnp.ndarray) -> jnp.ndarray:
+def denormalize(stats: Any, batch: jnp.ndarray) -> jnp.ndarray:
     """Transform normalized data back into original distribution
 
-    stats (array) -- mean, var, count
+    stats (Any)   -- dictionary with running mean, var, count.
     batch (array) -- current batch of data
 
     Returns:
         denormalize batch (array)
     """
 
-    mean, var = stats[0], stats[1]
-    denormalize_batch = batch * jnp.sqrt(var) + mean
+    mean, std = stats["mean"], stats["std"]
+    denormalize_batch = batch * std + mean
 
     return denormalize_batch
 
 
+def update_and_normalize_observations(stats: Any, observation: OLT) -> Tuple[Any, OLT]:
+    """Update running stats and normalise observations
+
+    stats (Dictionary)   -- array with running mean, var, count.
+    batch (OLT namespace)   -- current batch of data for a single agent.
+
+    Returns:
+        normalize batch (Dictionary)
+    """
+
+    obs_shape = observation.observation.shape
+    obs = jax.tree_util.tree_map(
+        lambda x: jnp.reshape(x, [-1] + list(x.shape[2:])), observation.observation
```

Comment:
haiku has a `merge_leading_dims` function which would be a bit cleaner

`from haiku._src.basic import merge_leading_dims`